ceph: add snapshot scheduling for mirrored pools (#6553)
rook/rook: Documentation/ceph-pool-crd.md

  mirroring:
    enabled: true
    mode: image
    # schedule(s) of snapshot
    snapshotSchedules:
      - interval: 24h # daily snapshots
        startTime: 14:00:00-05:00

>       # another schedule

Reviewer: seems fine to just have a single example instead of a comment that is the same, if you want to remove it

  mirroring:
    enabled: true
    mode: image

>     snapshotSchedule:

Reviewer 2: Just for clarification, but does this mean that rook will only support a single snapshot schedule per-pool?

Author: Hum yes? So Ceph can have multiple schedules for a given pool?

Reviewer 2: Yes, indeed. Plus you can apply schedules to individual images. The longer term goal is to be able to define named schedules and then just apply them (by-name) globally, to the pool, and/or to individual images as well.

Author: Rook does not apply schedules to individual images since it does not manage images. The only thing we can do is at the pool level. I guess we can make multiple schedules configurable.

Reviewer 2: Understood and agreed. It would be nice if Rook and the future CSI can have a similar approach (CRD) for defining schedules, though.

Author: Thanks, I'll add multiple schedules for this patch anyway.

* `mirroring`: Sets up mirroring of the pool
  * `enabled`: whether mirroring is enabled on that pool (default: false)
  * `mode`: mirroring mode to run, possible values are "pool" or "image" (required). Refer to the [mirroring modes Ceph documentation](https://docs.ceph.com/docs/master/rbd/rbd-mirroring/#enable-mirroring) for more details.
  * `snapshotSchedule`: schedule snapshot at the pool level **only**
    * `interval`: periodicity of the snapshots. The interval can be specified in days, hours, or minutes using d, h, m suffix respectively.

>     * `startTime`: optional, determines at what time the snapshot process starts. It can be specified using the ISO 8601 time format.

Author: > * How does the action of taking a snapshot impact the cluster? Does it pause the IO or just a brief perf impact? A brief note on this may help the user understand why they should care about the startTime.

Taking a snapshot is not disruptive so it's really up to them hence `startTime` being optional.

> * What is the default?

Probably the time when it was setup. @dillaman, what's the default start time for snapshot scheduling if nothing is set? Thanks!

>     * `startTime`: optional, determines at what time the snapshot process starts. It can be specified using the ISO 8601 time format.

Reviewer 2: > Probably the time when it was setup. @dillaman, what's the default start time for snapshot scheduling if nothing is set? Thanks!

00:00:00 UTC

* `mirroring`: Sets up mirroring of the pool
  * `enabled`: whether mirroring is enabled on that pool (default: false)
  * `mode`: mirroring mode to run, possible values are "pool" or "image" (required). Refer to the [mirroring modes Ceph documentation](https://docs.ceph.com/docs/master/rbd/rbd-mirroring/#enable-mirroring) for more details.

>   * `snapshotSchedules`: schedule(s) snapshot at the **pool** level. **Only** supported as of Ceph Octopus release. One then one schedule is supported.

Reviewer: ```suggestion
  * `snapshotSchedules`: schedule(s) snapshot at the **pool** level. **Only** supported as of Ceph Octopus release. One or more schedules are supported.
```

    * `interval`: frequency of the snapshots. The interval can be specified in days, hours, or minutes using d, h, m suffix respectively.

>     * `startTime`: optional, determines at what time the snapshot process starts. It can be specified using the ISO 8601 time format.

Reviewer: ```suggestion
    * `startTime`: optional, determines at what time the snapshot process starts, specified using the ISO 8601 time format.
```
